includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -376,185 +376,233 @@ public function payment_fields() {
-				$is_saved_method     = empty( $_POST['wcpay-payment-method'] ) && ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-payment-token' ] ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
-
-				// Determine the customer making the payment, create one if we don't have one already.
-				$user        = wp_get_current_user();
-				$customer_id = $this->customer_service->get_customer_id_by_user_id( $user->ID );
-
-				if ( null === $customer_id ) {
-					// Create a new customer.
-					$customer_id = $this->customer_service->create_customer_for_user( $user, $name, $email );
-				} else {
-					// Update the existing customer with the current details. In the event the old customer can't be
-					// found a new one is created, so we update the customer ID here as well.
-					$customer_id = $this->customer_service->update_customer_for_user( $customer_id, $user, $name, $email );
-				}
+	public function process_payment_for_order( $args ) {
+		// phpcs:disable Generic.Commenting.DocComment.MissingShort
+		$order           = $args['order'];           /** @var WC_Order $order The Order object. */
+		$cart            = $args['cart'];            /** @var WC_Cart $cart The Cart object, or NULL if the payment isn't made on a checkout context. */
+		$manual_capture  = $args['manual_capture'];  /** @var bool $manual_capture Whether to only authorize the payment (true) or automatically capture (false). */
+		$payment_method  = $args['payment_method'];  /** @var string $payment_method Payment Method ID. */
+		$token           = $args['token'];           /** @var WC_Payment_Token $token Token to use for the payment. */
+		$is_saved_method = $args['is_saved_method']; /** @var bool $is_saved_method Whether the payment is made using a saved card. */
+		$off_session     = $args['off_session'];     /** @var bool $off_session Whether this payment is merchant-initiated (true) of customer-initiated (false). */
+		$is_recurring    = $args['is_recurring'];    /** @var bool $is_recurring Whether this is a one-off payment (false) or it's the first installment of a recurring payment (true). */
+		// phpcs:enable Generic.Commenting.DocComment.MissingShort
+		$save_payment_method = ! $is_saved_method && ( ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-new-payment-method' ] ) || $is_recurring ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+
+		$order_id = $order->get_id();
+		$amount   = $order->get_total();
+		$user     = $order->get_user() ?? wp_get_current_user();
+		$name     = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
+		$email    = sanitize_email( $order->get_billing_email() );
+		$metadata = [
+			'customer_name'  => $name,
+			'customer_email' => $email,
+			'site_url'       => esc_url( get_site_url() ),
+			'order_id'       => $order_id,
+		];
+		if ( $is_recurring ) {
+			$metadata['payment_type'] = 'recurring';
+		}
```

Comment:
`recurring` or `single` maybe?

Reply:
cc @allendav @DanReyLop @RadoslavGeorgiev

Reply:
Since older (single) transactions won't have this field at all, I like your idea - i.e. add recurring: true to $metadata for recurring transactions and don't add it at all otherwise

Reply:
Will there be any other type of payments we'll need to think about in the future? I'd rather not create a bunch of metadata keys that we need to check for in future implementations.

E.g. would a point of sale payment be considered it's own payment type? It would be really annoying to have to check for a `recurring` key and a `point_of_sale` key.

Which is why I was thinking about making this a bit more explicit than a boolean flag.

> Since older (single) transactions won't have this field at all

I think if we go for something more explicit than a boolean flag (like a list of possible string values) we can assume that if the `payment_type` key is missing it's a single payment. That's no different than always checking for the existence of a `recurring` key.

Reply:
What if it is a recurring point-of-sale transaction? :P 

All kidding aside, I don't feel strongly about this. It is hard to tell at this point what we might do besides recurring, but I agree that there is likely to be something at some point, so, sure - let's do something like payment_type: recurring and treat missing as a non-recurring

Reply:
Pre-orders and deposits are two examples of potential future payment types.

Reply:
This should be resolved in #861, so I'm marking this discussion as resolved 🙂